1 - O que é Aprendizado de Máquina?

Aprendizado de máquina é um tipo de algoritmo que pode aprender algo sem instruções. 

2 - O que caracteriza um problema de classificação?

Os problemas de classificação são aqueles onde se busca encontrar uma classe, dentro das possibilidades limitadas existentes.
Esta classe pode ser se um aluno foi aprovado ou reprovado, se uma pessoa possui uma doença ou não, dentre outras tantas possibilidades, sendo que nestes casos ou a previsão será uma ou outra.

Um algoritmo de classificaão nada mais é do que um passo-a-passo a ser seguido para realizar alguma tarefa específica.

3 - Diferenças entre SVM e Árvores de Decisão

Uma máquina de vetores de suporte (SVM) é um algoritmo de aprendizado de máquina supervisionado que classifica dados encontrando uma linha ou hiperplano ideal que maximize a distância entre cada classe em um espaço n-dimensional.

O algoritmo SVM é amplamente usado no aprendizado de máquina, pois é capaz de lidar com tarefas de classificação linear e não linear. No entanto, quando os dados não são linearmente separáveis, as funções kernel são usadas para transformar os dados em um espaço de dimensão superior para permitir a separação linear.

Essa aplicação de funções kernel pode ser conhecida como "truque de kernel", e a escolha da função kernel, como kernels lineares, kernels polinomiais, kernels de função de base radial (RBF) ou kernels sigmoides, depende das características dos dados e do caso de uso específico.

====================================================================================================================

As Árvores de Decisão são algoritmos intuitivos que organizam dados hierarquicamente, criando uma estrutura em forma de árvore para facilitar decisões. Cada nó representa uma decisão baseada em atributos dos dados, dividindo-os até um resultado final. Funciona como um fluxograma que guia para a resposta certa. 

Árvores de decisão são fáceis de entender e lidar com dados variados, captando relações não lineares. Sofrem overfitting em pequenos conjuntos, mas a poda ajuda. Pequenas mudanças nos dados podem afetar a robustez. Usadas em diagnósticos médicos e detecção de fraudes, identificam padrões com eficácia.

4 - Carregamento e visualização do dataset

In [4]:
import pandas as pd

data = pd.read_csv('diabetes.csv')

print(data.head())

print(data.describe())



   Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1  
       Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
count   768.000000  768.000000     768.000000     768.000000  768.000000   
mean      3.845052  120.894531      69.105469      20.536458   79.799479   
std       3.369578   31.972618      19.355807      15.952218  115.244002   

5 - Divisão em treino e teste

In [7]:
from sklearn.model_selection import train_test_split

X = data.drop('Outcome', axis=1)  
y = data['Outcome'] 

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

print(f'Tamanho do conjunto de treino: {X_train.shape}')
print(f'Tamanho do conjunto de teste: {X_test.shape}')


Tamanho do conjunto de treino: (614, 8)
Tamanho do conjunto de teste: (154, 8)


6 - Normalização dos dados

In [8]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

X_train_scaled = scaler.fit_transform(X_train)

X_test_scaled = scaler.transform(X_test)

print("Dados de treino normalizados:")
print(X_train_scaled[:5])

print("\nDados de teste normalizados:")
print(X_test_scaled[:5])


Dados de treino normalizados:
[[-0.52639686 -1.15139792 -3.75268255 -1.32277365 -0.70120553 -4.13525578
  -0.49073479 -1.03594038]
 [ 1.58804586 -0.27664283  0.68034485  0.23350519 -0.70120553 -0.48916881
   2.41502991  1.48710085]
 [-0.82846011  0.56687102 -1.2658623  -0.09071957  0.01344832 -0.42452187
   0.54916055 -0.94893896]
 [-1.13052335  1.2541786  -1.04961706 -1.32277365 -0.70120553 -1.30372015
  -0.63929127  2.79212217]
 [ 0.68185612  0.41066475  0.57222224  1.07648956  2.48460077  1.83812075
  -0.68682934  1.13909516]]

Dados de teste normalizados:
[[ 0.68185612 -0.71402038 -0.61712658  0.81710976  0.93474906  0.26073561
  -0.11637247  0.87809089]
 [-0.52639686 -0.27664283  0.30191569  0.7522648  -0.70120553  0.48053518
  -0.954231   -1.03594038]
 [-0.52639686 -0.40160784 -0.29275872 -1.32277365 -0.70120553 -0.15300476
  -0.9245197  -1.03594038]
 [ 1.28598261 -0.43284909  0.57222224 -1.32277365 -0.70120553 -0.95462672
   1.14932872  0.0950781 ]
 [ 0.98391937  0.47314726  1.1

7 - Treinamento de um modelo SVM

In [9]:
from sklearn.svm import SVC

svm_model = SVC()

svm_model.fit(X_train_scaled, y_train)

y_pred = svm_model.predict(X_test_scaled)

print("Previsões no conjunto de teste: ")
print(y_pred[:10])

Previsões no conjunto de teste: 
[0 0 0 0 0 0 0 1 1 1]


8 - Treinamento de um modelo com Árvore de Decisão

In [10]:
from sklearn.tree import DecisionTreeClassifier

dt_model = DecisionTreeClassifier(random_state=42)

dt_model.fit(X_train, y_train)

y_pred_dt = dt_model.predict(X_test)

print("Previsões no conjunto de teste com Árvore de Decisão")
print(y_pred_dt[:10])


Previsões no conjunto de teste com Árvore de Decisão
[1 0 0 0 0 0 0 0 0 1]


9 - Avaliação dos modelos

In [11]:
from sklearn.metrics import classification_report

print("Avaliação do Modelo SVM: ")
print(classification_report(y_test, y_pred))

print("Avaliação do Modelo Árvore de Decisão: ")
print(classification_report(y_test, y_pred_dt))


Avaliação do Modelo SVM: 
              precision    recall  f1-score   support

           0       0.77      0.83      0.80        99
           1       0.65      0.56      0.60        55

    accuracy                           0.73       154
   macro avg       0.71      0.70      0.70       154
weighted avg       0.73      0.73      0.73       154

Avaliação do Modelo Árvore de Decisão: 
              precision    recall  f1-score   support

           0       0.83      0.76      0.79        99
           1       0.62      0.73      0.67        55

    accuracy                           0.75       154
   macro avg       0.73      0.74      0.73       154
weighted avg       0.76      0.75      0.75       154



10 - Ajuste de Hiperparâmetros

In [12]:
#Hiperparâmetro SVM

svm_model_linear = SVC(kernel='linear')
svm_model_linear.fit(X_train_scaled, y_train)
y_pred_linear = svm_model_linear.predict(X_test_scaled)

svm_model_rbf = SVC(kernel='rbf')
svm_model_rbf.fit(X_train_scaled, y_train)
y_pred_rbf = svm_model_rbf.predict(X_test_scaled)

from sklearn.metrics import classification_report

print("Avaliação do SVM com kernel 'linear':")
print(classification_report(y_test, y_pred_linear))

print("Avaliação do SVM com kernel 'rbf':")
print(classification_report(y_test, y_pred_rbf))


Avaliação do SVM com kernel 'linear':
              precision    recall  f1-score   support

           0       0.81      0.82      0.81        99
           1       0.67      0.65      0.66        55

    accuracy                           0.76       154
   macro avg       0.74      0.74      0.74       154
weighted avg       0.76      0.76      0.76       154

Avaliação do SVM com kernel 'rbf':
              precision    recall  f1-score   support

           0       0.77      0.83      0.80        99
           1       0.65      0.56      0.60        55

    accuracy                           0.73       154
   macro avg       0.71      0.70      0.70       154
weighted avg       0.73      0.73      0.73       154



In [13]:
#Hiperparâmetro Árvore de Decisão

dt_model_depth3 = DecisionTreeClassifier(max_depth=3, random_state=42)
dt_model_depth3.fit(X_train, y_train)
y_pred_depth3 = dt_model_depth3.predict(X_test)

dt_model_no_limit = DecisionTreeClassifier(max_depth=None, random_state=42)
dt_model_no_limit.fit(X_train, y_train)
y_pred_no_limit = dt_model_no_limit.predict(X_test)

print("Avaliação da Árvore de Decisão com max_depth=3:")
print(classification_report(y_test, y_pred_depth3))

print("Avaliação da Árvore de Decisão sem limite de profundidade:")
print(classification_report(y_test, y_pred_no_limit))


Avaliação da Árvore de Decisão com max_depth=3:
              precision    recall  f1-score   support

           0       0.80      0.84      0.82        99
           1       0.68      0.62      0.65        55

    accuracy                           0.76       154
   macro avg       0.74      0.73      0.73       154
weighted avg       0.76      0.76      0.76       154

Avaliação da Árvore de Decisão sem limite de profundidade:
              precision    recall  f1-score   support

           0       0.83      0.76      0.79        99
           1       0.62      0.73      0.67        55

    accuracy                           0.75       154
   macro avg       0.73      0.74      0.73       154
weighted avg       0.76      0.75      0.75       154



11 - Relatório Final

Análise Comparativa entre os Modelos SVM e Árvore de Decisão Desempenho

    Com base nos resultados obtidos, os dois modelos apresentam características distintas, influenciando seu desempenho em diferentes contextos:

    Modelo SVM (com kernel ajustado):

    É altamente eficaz para conjuntos de dados bem estruturados e que não possuem uma quantidade excessiva de ruído.

    O SVM tende a ter um F1-score mais elevado, indicando um equilíbrio mais consistente entre Precisão e Revocação.

    A normalização dos dados contribuiu para o desempenho estável do modelo, evidenciado pela acurácia consistente, especialmente para o kernel 'rbf', que é ideal  
    para dados com relações complexas.

    Modelo Árvore de Decisão (com max_depth ajustado):

    A Árvore de Decisão é intuitiva e interpretável, sendo fácil rastrear como as decisões foram tomadas.

    Em termos de precisão geral, o modelo pode sofrer com overfitting em árvores muito profundas (sem restrição de max_depth), resultando em um desempenho inferior ao 
    SVM nos dados de teste.

    Contudo, para dados menores ou mais simples, com alta linearidade, a Árvore de Decisão pode se apresentar como uma alternativa eficiente.

    Critérios para a Escolha do Modelo Mais Adequado

Desempenho Estatístico:

Desempenho Geral
SVM (kernel 'linear'):

Acurácia: 76%
F1-score: Maior equilíbrio geral para as duas classes, com 0.81 para a classe 0 e 0.66 para a classe 1.
A alta precisão para a classe 0 (81%) sugere bom desempenho em identificar corretamente os negativos. No entanto, a revocação para a classe 1 é mais baixa (65%), mostrando que o modelo tem maior dificuldade em identificar verdadeiros positivos.

SVM (kernel 'rbf'):

Acurácia: 73%
F1-score: Menor do que o kernel 'linear', especialmente para a classe 1 (0.60).
Embora o kernel 'rbf' seja mais eficaz em modelar relações complexas, neste caso específico ele apresentou desempenho inferior ao kernel 'linear'.

Árvore de Decisão (com max_depth=3):

Acurácia: 76%
F1-score: Muito próximo ao SVM com kernel 'linear', sendo 0.82 para a classe 0 e 0.65 para a classe 1.
Limitar a profundidade da árvore evitou overfitting e manteve a consistência no desempenho.

Árvore de Decisão (sem limite de profundidade):

Acurácia: 75%
F1-score: O resultado foi levemente inferior ao caso com max_depth=3, indicando que permitir maior profundidade resultou em possível overfitting, prejudicando a generalização.


Interpretação e Tempo de Treinamento:

Para aplicações em que a interpretabilidade é crucial, a Árvore de Decisão se torna mais vantajosa, permitindo que as regras do modelo sejam visualizadas.

Além disso, a Árvore de Decisão é mais rápida para treinar, o que a torna uma boa opção para prototipagem rápida.

Escalabilidade:

O SVM pode ter dificuldades de escalabilidade em grandes conjuntos de dados, enquanto a Árvore de Decisão é mais escalável e se adapta melhor a grandes volumes de dados.

Recomendação para Uso em Produção
O modelo SVM seria o mais indicado para uso em produção em cenários onde:

Há necessidade de maior precisão.
Os dados são pré-processados adequadamente (normalizados).
O desempenho do modelo com kernel ajustado (ex.: 'rbf') apresentou superioridade consistente nos testes.

Por outro lado, o modelo Árvore de Decisão seria mais adequado em contextos onde:

A interpretabilidade do modelo é uma prioridade.
A escala e a velocidade de treinamento são considerações importantes.